processor/src/main/java/org/mapstruct/ap/internal/processor/creation/MappingResolverImpl.java

```diff
@@ -426,29 +426,33 @@ private Assignment resolveViaMethodAndMethod(Type sourceType, Type targetType) {
             // sourceMethod or builtIn that fits the signature B to C. Only then there is a match. If we have a match
             // a nested method call can be called. so C = methodY( methodX (A) )
             for ( Method methodYCandidate : methodYCandidates ) {
-                if ( Object.class.getName()
-                    .equals( methodYCandidate.getSourceParameters().get( 0 ).getType().getName() ) ) {
+                Type ySourceType = methodYCandidate.getSourceParameters().get( 0 ).getType();
+                if ( Object.class.getName().equals( ySourceType.getName() ) ) {
                     //  java.lang.Object as intermediate result
                     continue;
                 }
 
-                methodRefY =
-                    resolveViaMethod( methodYCandidate.getSourceParameters().get( 0 ).getType(), targetType, true );
+                if ( ySourceType.isTypeVar() ) {
+                    ySourceType = ySourceType.resolveToType( targetType, methodYCandidate.getResultType() );
+                }
```

Comment:
`resolveToType` already checks for `isTypeVar()`. You can directly do

```
ySourceType = ySourceType.resolveToType( targetType, methodYCandidate.getResultType() );
```

Reply:
that's true..  but there's actually a reason for this:
1. I did not want to do the resolveToType unchecked in the `resolveToType` method itself.. Was a bit struggling with the method name.
2. I wanted to make the flow clear at the call site..

 I could have made it `resolveTypeVarToType` at hind site.  Will do this.